pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -0,0 +1,1150 @@
+			linuxResource.CpuShares = &cpuShares
+		}
+		if memoryLimit != 0 {
+			linuxResource.MemoryLimitInBytes = &memoryLimit
+		}
+		config.Linux = &runtimeApi.LinuxContainerConfig{
+			Resources: linuxResource,
+		}
+	}
+
+	if container.SecurityContext != nil {
+		securityContext := container.SecurityContext
+		if securityContext.Privileged != nil {
+			config.Privileged = securityContext.Privileged
+		}
+
+		if securityContext.ReadOnlyRootFilesystem != nil {
+			config.ReadonlyRootfs = securityContext.ReadOnlyRootFilesystem
+		}
+
+		if securityContext.Capabilities != nil {
+			if config.Linux == nil {
+				config.Linux = &runtimeApi.LinuxContainerConfig{
+					Capabilities: &runtimeApi.Capability{
+						AddCapabilities:  make([]string, 0, len(securityContext.Capabilities.Add)),
+						DropCapabilities: make([]string, 0, len(securityContext.Capabilities.Drop)),
+					},
+				}
+			}
+
+			for index, value := range securityContext.Capabilities.Add {
+				config.Linux.Capabilities.AddCapabilities[index] = string(value)
+			}
+			for index, value := range securityContext.Capabilities.Drop {
+				config.Linux.Capabilities.DropCapabilities[index] = string(value)
+			}
+		}
+
+		if securityContext.SELinuxOptions != nil {
+			if config.Linux == nil {
+				config.Linux = &runtimeApi.LinuxContainerConfig{}
+			}
+			config.Linux.SelinuxOptions = &runtimeApi.SELinuxOption{
+				User:  &securityContext.SELinuxOptions.User,
+				Role:  &securityContext.SELinuxOptions.Role,
+				Type:  &securityContext.SELinuxOptions.Type,
+				Level: &securityContext.SELinuxOptions.Level,
+			}
+		}
+	}
+
+	envs := make([]*runtimeApi.KeyValue, len(opts.Envs))
+	for index, e := range opts.Envs {
+		envs[index] = &runtimeApi.KeyValue{
+			Key:   &e.Name,
+			Value: &e.Value,
+		}
+	}
+	config.Envs = envs
+
+	return config, nil
+}
+
+func (m *kubeGenericRuntimeManager) startContainer(podSandboxID string, container *api.Container, pod *api.Pod, podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret) (error, string) {
+	err, msg := m.imagePuller.PullImage(pod, container, pullSecrets)
+	if err != nil {
+		return err, msg
+	}
+
+	ref, err := kubecontainer.GenerateContainerRef(pod, container)
+	if err != nil {
+		glog.Errorf("Can't make a ref to pod %v, container %v: '%v'", pod.Name, container.Name, err)
+	}
+	glog.Infof("Generating ref for container %s: %#v", container.Name, ref)
+
+	// For a new container, the RestartCount should be 0
+	var restartCount int = 0
+	containerStatus := podStatus.FindContainerStatusByName(container.Name)
+	if containerStatus != nil {
+		restartCount = containerStatus.RestartCount + 1
+	}
+
+	containerConfig, err := m.generateContainerConfig(container, pod, restartCount)
+	if err != nil {
+		m.recorder.Eventf(ref, api.EventTypeWarning, events.FailedToCreateContainer, "Failed to create container with error: %v", err)
+		return err, "Generate Container Config Failed"
+	}
+
+	podSandboxConfig, err := m.generatePodSandboxConfig(pod)
```

Comment:
Why can't we generate the `PodSandboxConfig` once and pass it to each container that we need to start?